[quant] Create PerRowQuantizer for floating point scale and zero_point

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42881 [quant] Add torchbind support for embedding_bag packed weights
* #42762 [quant] Add embeddingbag_prepack function that works on quantized tensor.
* #42690 [quant] Make PerChannel Observer work with float qparams
* **#42612 [quant] Create PerRowQuantizer for floating point scale and zero_point**

Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

supriyar committed:
[quant] Create PerRowQuantizer for floating point scale and zero_point

Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
Note to reviewers: This change touches quite a few files. But the relevant ones are 
Quantizer.h/.cpp - Adds a new type of quantizer - PerRowFloatQParamsQuantizer
QScheme.h - Define a new qscheme corresponding to this quantizer
QuantizedOpKernels - kernel for new quantize/dequantize function for this quantizer
affine_quantizer.cpp - `quantize_val_float_qparams` which is the quantize_val equation for this case.

Rest of the other files add necessary code to hook up the new quantizer to the top level API and add print support.

supriyar commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -314,6 +395,25 @@ CAFFE2_API float dequantize_val(double scale, int64_t zero_point, T value) {
+  constexpr int64_t qmin = std::numeric_limits<typename T::underlying>::min();
+  constexpr int64_t qmax = std::numeric_limits<typename T::underlying>::max();
+  float inv_scale = scale == 0 ? 1.0f : 1.0f / scale;
+  qvalue = lrintf((value - bias) * inv_scale);
```
cc @raghuramank100, @dskhudia please check if this equation makes sense. It is what we currently use in caffe2 today for quantizing embedding weights.

dskhudia replied:
This looks fine to me. You may also want to check for inv_scale being inf as some day we will hit that case.

vkuzo replied:
> You may also want to check for inv_scale being inf as some day we will hit that case.

+1

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.h:
```diff
@@ -155,6 +155,62 @@ struct CAFFE2_API PerChannelAffineQuantizer : public AffineQuantizer {
 };
 
+/**
+ * PerRowFloatQParamsQuantizer is the same as PerChannelAffineQuantizer
```
can the naming reflect this similarity? e.g. PerChannelAffineFloatZeroPoint

supriyar replied:
Sure, I can call it PerChannelFloatQParamsQuantizer. Will update names in other places too.

jerryzh168 replied:
please make it `PerChannelAffineFloatQParamsQuantizer`, we want to make it clear this is a variant of `PerChannelAffineQuantizer`

dskhudia commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -80,6 +82,13 @@ void checkZeroPoints(const std::string& fn_name, Tensor zero_points) {
   }
 }
 
+void checkFloatZeroPoints(const std::string& fn_name, Tensor zero_points) {
```
Is this needed? float values will always fit in float type.

supriyar replied:
Yeah, can remove this check

jerryzh168 commented on aten/src/ATen/quantized/Quantizer.h:
```diff
@@ -155,6 +155,62 @@ struct CAFFE2_API PerChannelAffineQuantizer : public AffineQuantizer {
+ * be exactly represented in the quantized space. We can get additional precision by
+ * using floating point values for zero point.
+ */
+struct CAFFE2_API PerRowFloatQParamsQuantizer : public AffineQuantizer {
```
since the fields and methods are the same, maybe this can inherit from PerChannelAffineQuantizer and reuse all the methods

supriyar replied:
This will need a different qscheme and different quantize/dequantize functions. There is a comment in QScheme.h that each new qscheme should correspond 1-1 to a quantizer. https://github.com/pytorch/pytorch/blob/master/c10/core/QScheme.h

jerryzh168 replied:
yeah, we can still inherit from PerChannelAffineQuantizer for the implementation of this Quantizer to reduce code duplication and also make it more clear this qscheme is a variant of the per channel affine quantization.

raghuramank100 replied:
Dont we use the same quantizers for the symmetric and affine cases?

jerryzh168 replied:
no, we don't have symmetric quantized tensor in the backend. so we don't have symmetric quantizer.

raghuramank100 commented on aten/src/ATen/core/Formatting.cpp:
```diff
@@ -281,7 +281,8 @@ std::ostream& print(std::ostream& stream, const Tensor & tensor_, int64_t linesi
         stream << ", zero_point: " << tensor_.q_zero_point();
-      } else if (tensor_.qscheme() == c10::kPerChannelAffine) {
+      } else if (tensor_.qscheme() == c10::kPerChannelAffine ||
+          tensor_.qscheme() == c10::kPerRowFloatQParams) {
```
Can we call this PerChannelFloatQParams as this generalized per-channel quant?

supriyar replied:
Yes, I will be renaming these shortly and pushing an update.

raghuramank100 commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -147,6 +156,43 @@ Tensor quantize_tensor_per_channel_affine(
+
+  TORCH_CHECK(
+      0 <= axis && axis < rtensor.dim(),
+      "Channel axis out of range in per row float qparams quantization. Got: ",
```
Lets avoid using per-row as the quantizer itself can work for any tensor

raghuramank100 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -40,8 +40,13 @@ Tensor quantize_per_channel_cpu(
+    quantizer = make_per_row_float_qparams_quantizer(scales, zero_points, axis, dtype);
+  } else {
+    quantizer =
+        make_per_channel_affine_quantizer(scales, zero_points, axis, dtype);
```
Can we overload the make_per_channel_affine_quantizer function here to handle different dtypes for scale and zero-point?

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -314,6 +386,25 @@ CAFFE2_API float dequantize_val(double scale, int64_t zero_point, T value) {
 
+/*
+* Quantize value based on the following equation
+* Qx = (Xf - bias) * inv_scale
```
just to clarify, does `bias = -1 * zero_point * scale` (to get `Qx = round(Xf/scale - zero_point)`?

if yes, maybe we can make the API clearer by having the callsites also use the "bias" name?

supriyar replied:
I think this was a little confusing as I was trying to match the C2 equation  `(Xf-bias) * inv_scale` in numerics by setting zero_point = bias.
I had a chat with Raghu about this and we thought it would be better to use zero_point = (-bias/scale) for this case so that we can use similar quantize equation to what we have now. 
So the current equation: `Xq = Xf * inv_scale + zp`
Substituting zero_point
`Xq = Xf * inv_scale + (-bias *inv_scale)`
`Xq = (Xf - bias) * inv_scale`, which is the same as what caffe2 uses for embedding layers today. There may be some numerical differences due to the division rounding.
I'll update the PR with the change for this.

vkuzo replied:
yes, thanks, that matches what I assumed from reading the code.  I think that works well, but maybe we can have the callers use "bias" and not "zero_point" as the argument name then, to prevent confusion?  i.e.

```
quantize_val_float_qparams<qint8>(float scale, float bias, float value);
# instead of
quantize_val_float_qparams<qint8>(float scale, float zero_point, float value);
```

vkuzo commented on c10/core/QScheme.h:
```diff
@@ -16,13 +16,15 @@ enum class QScheme : uint8_t {
   PER_CHANNEL_SYMMETRIC = 3,
-  COMPILE_TIME_NUM_QSCHEMES = 4,
+  PER_CHANNEL_AFFINE_FLOAT_QPARAMS = 4,
+  COMPILE_TIME_NUM_QSCHEMES = 5,
```
just curious, what is `COMPILE_TIME_NUM_QSCHEMES`? is there a reason it has to be last?

supriyar replied:
i think it denotes the number of qschemes we have

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -314,6 +386,27 @@ CAFFE2_API float dequantize_val(double scale, int64_t zero_point, T value) {
+* to (-Xmin/scale), where Xmin is the min value in input tensor row.
+*/
+template <typename T>
+T quantize_val_float_qparams(float scale, float zero_point, float value) {
```
@vkuzo, @raghuramank100 I've updated the float qparams function to use the same equation as our current quantize function. Please take a look 
The zero_point passed in will be (-Xmin/scale), so that we can match the caffe2 quantize function used here - https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp#L119-L122

vkuzo commented on aten/src/ATen/quantized/Quantizer.h:
```diff
@@ -149,12 +149,57 @@ struct CAFFE2_API PerChannelAffineQuantizer : public AffineQuantizer {
+ * kPerChannelAffine.
+ *
+ * The quantize equation in this case looks like -
+ * Xq = (Xf - zero_point) * inv_scale, where inv_scale = 1.0/scale
```
hmm, so just to confirm, this is still not the same `zero_point` conceptually that we use elsewhere?  I feel like people might get confused by this.  What would you think of naming it `bias` (as I think it was earlier in this PR), or `zero_point_caffe2`, etc?

jerryzh168 replied:
I think this is not a problem, we have different quantize functions as PerchannelAffineQuantizer

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -57,7 +59,7 @@ void checkQuantizedTensor(const std::string& fn_name, Tensor t) {
 
 template <typename T>
-void checkZeroPoint(const std::string& fn_name, int64_t zero_point) {
+void checkZeroPoint(const std::string& fn_name, T zero_point) {
```
I think `T` here is for things like `int8`, `uint8` etc.?

supriyar replied:
Yeah we don't need to check zero_point when it is float.

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LGTM

jerryzh168 commented on aten/src/ATen/native/quantized/affine_quantizer.cpp:
```diff
@@ -147,6 +149,42 @@ Tensor quantize_tensor_per_channel_affine(
+    int64_t axis) {
+  static const auto fn_name = "quantize_tensor_per_channel_float_qparams";
+
+  checkRoundingMode(fn_name);
+  checkFloatTensor(fn_name, rtensor);
+  checkCPUTensor(fn_name, rtensor);
+  checkSameDevice(fn_name, rtensor, qtensor);
+  checkSameSize(fn_name, qtensor, rtensor);
```
these probably should be put into one function

supriyar committed:
Update on "[quant] Create PerRowQuantizer for floating point scale and zero_point"


Summary:
Add a new Quantizer that supports an input zero point (bias) that can be float.
The quantization equation in this case is

Xq = (Xf - bias) * inv_scale, where bias is float zero_point value
We start with per-row implementation and can extend to per-tensor in the future, if necessary

Test Plan:
python test/test_quantization.py TestQuantizedTensor

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22960142](https://our.internmc.facebook.com/intern/diff/D22960142)

[ghstack-poisoned]
